types/events.ts

```diff
@@ -0,0 +1,288 @@
+    percent: ProportionPercent;
+    seconds: Seconds;
+}
+
+export interface FullscreenChangeEvent {
+    fullscreen: boolean;
+}
+
+export interface TextTrackChangeEvent {
+    kind: string | undefined;
+    label: string | undefined;
+    language: string | undefined;
+}
+
+export interface ChapterChangeEvent {
+    startTime: Seconds;
+    /** Title of the chapter */
+    title: string;
+    /**
+     * The index property of each chapter is the place it holds
+     * in the order of all the chapters. It starts at 1.
+     */
+    index: number;
+}
+
+export interface CuePointEvent<T = Record<string, unknown>> extends VimeoCuePoint<T> {}
+
+export interface CueChangeEvent extends TextTrackChangeEvent {
+    cues: VimeoTextTrackCuePoint[];
+}
+
+export interface LoadedEvent {
+    id: number;
+}
+
+export interface DurationChangeEvent {
+    duration: Seconds;
+}
+
+export interface VolumeChangeEvent {
+    volume: VolumeLevel;
+    muted: boolean;
+}
+
+export interface PlaybackRateChangeEvent {
+    playbackRate: PlaybackRate;
+}
+
+export interface QualityChangeEvent {
+    quality: VideoQuality;
+}
+
+export interface CameraChangeEvent extends CameraProperties {}
+
+export interface ResizeEvent {
+    videoWidth: Pixels;
+    videoHeight: Pixels;
+}
+
+export interface ErrorEvent {
+    name: string;
+    message: string;
+    method: string;
+}
+
+export interface InteractiveHotspotClickEvent {
+    action: 'seek' | 'event' | 'none' | 'overlay' | 'url';
+    actionPreference: {
+        /**
+         * // on `event`, `overlay`, `seek`, `url` action
+         */
+        pauseOnAction?: boolean;
+        /**
+         * on `overlay` action
+         */
+        overlayId?: number;
+        /**
+         * on `seek` action
+         */
+        seekTo?: number;
+        /**
+         * on `url` action
+         */
+        url?: string;
+    };
+    currentTime: number;
+    customPayloadData: any | null;
+    hotspotId: number;
+}
+export interface InteractiveOverlayPanelClickEvent {
+    action: 'seek' | 'clickthrough' | 'close' | 'event' | 'none';
+    actionPreference: {
+        /** on `close`, `seek` action */
+        pauseOnAction?: boolean;
+        /** on `seek` action */
+        seekTo?: number;
+        /** on `clickthrough` action */
+        url?: string;
+    };
+    currentTime: Seconds;
+    customPayloadData: unknown | null;
+    overlayId: number;
+    panelId: string;
+}
+
+/**
+ * Utility type that maps an event's data type to a generic version if supported
+ */
+export type EventDataWithGenerics<E extends keyof PlayerEventMap | PlayerEvent, T> = E extends keyof PlayerEventMap 
+    ? PlayerEventMap[E] extends CuePointEvent<any>
+        ? CuePointEvent<T>
+        : PlayerEventMap[E]
+    : PlayerEventMap[E extends PlayerEvent ? E : never];
+
+/** Maps event names to their corresponding event data types */
+export interface PlayerEventMap {
+    [PlayerEvent.Play]: VimeoEvent;
+    play: VimeoEvent;
```

Comment:
Why do we duplicate it?

Reply:
@jdreetz I see. I'm curious if string constants would resolve that for us. Also,  it removes the need for an explicit enum object.

Reply:
I initially used just strings, but we don't get JS Doc comments for the events with those, so that's why I added support for the Enum versions as well.